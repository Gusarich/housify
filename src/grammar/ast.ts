export type AstModule = {
    kind: 'module';
    items: AstModuleItem[];
    id: number;
};

export type AstModuleItem = AstHouse;

export type AstHouse = {
    kind: 'house';
    name: AstId;
    items: AstHouseItem[];
    id: number;
};

export type AstHouseItem = AstGlobalStat | AstPlayerStat | AstHandler;

export type AstGlobalStat = {
    kind: 'globalStat';
    name: AstId;
    type: AstId;
    defaultValue?: AstExpression;
    id: number;
};

export type AstPlayerStat = {
    kind: 'playerStat';
    name: AstId;
    type: AstId;
    defaultValue?: AstExpression;
    id: number;
};

export type AstHandler = {
    kind: 'handler';
    event: AstId;
    statements: AstStatement[];
    id: number;
};

export type AstStatement =
    | AstStatementAssign
    | AstStatementLet
    | AstStatementExpression
    | AstStatementIf;

export type AstStatementAssign = {
    kind: 'statementAssign';
    lvalue: AstExpression;
    value: AstExpression;
    id: number;
};

export type AstStatementLet = {
    kind: 'statementLet';
    name: AstId;
    type: AstId;
    value: AstExpression;
    id: number;
};

export type AstStatementExpression = {
    kind: 'statementExpression';
    expression: AstExpression;
    id: number;
};

export type AstStatementIf = {
    kind: 'statementIf';
    condition: AstExpression;
    then: AstStatement[];
    else?: AstStatement[];
    id: number;
};

export type AstExpression =
    | AstExpressionBinary
    | AstExpressionUnary
    | AstIntegerLiteral
    | AstExpressionId;

export type AstBinaryOp =
    | '+'
    | '-'
    | '*'
    | '/'
    | '=='
    | '!='
    | '<'
    | '<='
    | '>'
    | '>='
    | '&&'
    | '||';

export type AstExpressionBinary = {
    kind: 'expressionBinary';
    op: AstBinaryOp;
    left: AstExpression;
    right: AstExpression;
    id: number;
};

export type AstUnaryOp = '+' | '-' | '!';

export type AstExpressionUnary = {
    kind: 'expressionUnary';
    op: AstUnaryOp;
    operand: AstExpression;
    id: number;
};

export type AstIntegerLiteral = {
    kind: 'integerLiteral';
    value: string;
    id: number;
};

export type AstExpressionId = {
    kind: 'expressionId';
    name: AstId;
    id: number;
};

export type AstId = {
    kind: 'id';
    name: string;
    id: number;
};

export type AstNode =
    | AstModule
    | AstModuleItem
    | AstHouseItem
    | AstStatement
    | AstExpression
    | AstId;

let nextId = 1;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;

export function createAstNode(src: DistributiveOmit<AstNode, 'id'>): AstNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}

export function cloneAstNode<T extends AstNode>(src: T): T {
    return { ...src, id: nextId++ };
}

export function resetNodeId() {
    nextId = 1;
}
