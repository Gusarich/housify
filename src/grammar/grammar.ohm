Housify {
    Module = ModuleItem*

    ModuleItem = House
    
    House = house id "{" HouseItem* "}"

    HouseItem = GlobalStat | PlayerStat | Handler

    GlobalStat = global id ":" Type ("=" Expression)? ";"
    PlayerStat = player id ":" Type ("=" Expression)? ";"

    Handler = handle id "{" Statement* "}"

    Statement = StatementAssign
              | StatementLet
              | StatementExpression
              | StatementIf
    
    StatementAssign = Expression "=" Expression ";"

    StatementLet = let id ":" Type "=" Expression ";"

    StatementExpression = Expression ";"

    StatementIf = if Expression "{" Statement* "}" ~else --noElse
                | if Expression "{" Statement* "}" else "{" Statement* "}" --withElse
                | if Expression "{" Statement* "}" else StatementIf --withElseIf
    
    Expression = ExpressionBinary
               | ExpressionUnary
               | ExpressionLiteral
               | ExpressionId
               | ExpressionParens
    
    ExpressionBinary = Expression binaryOp Expression
    ExpressionUnary = unaryOp Expression
    ExpressionLiteral = integerLiteral
    ExpressionId = id
    ExpressionParens = "(" Expression ")"

    binaryOp = "+" | "-" | "*" | "/" | "==" | "!=" | "<" | "<=" | ">" | ">=" | "&&" | "||"
    unaryOp = "+" | "-" | "!"

    Type = id --regular

    integerLiteral = digit+

    letterAsciiLC = "a".."z"
    letterAsciiUC = "A".."Z"
    letterAscii = letterAsciiLC | letterAsciiUC
    letterComment = letterAsciiLC | letterAsciiUC | digit | "_"

    idStart = letterAscii | "_"
    idPart = letterAscii | digit | "_"
    id = ~keyword #idStart #(idPart*)

    house = "house" ~idPart
    global = "global" ~idPart
    player = "player" ~idPart
    handle = "handle" ~idPart
    let = "let" ~idPart
    if = "if" ~idPart
    else = "else" ~idPart

    keyword = house | global | player | handle | let | if | else
}