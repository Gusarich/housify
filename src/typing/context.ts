import { Type } from './type';

export class StatementContext {
    variables: Map<string, Type>;
    types: Map<string, Type>;

    constructor() {
        this.variables = new Map();
        this.types = new Map();
    }

    clone() {
        const sctx = new StatementContext();
        for (const [name, type] of this.variables) {
            sctx.addVariable(name, type);
        }
        for (const [name, type] of this.types) {
            sctx.addType(name, type);
        }
        return sctx;
    }

    addVariable(name: string, type: Type) {
        this.variables.set(name, type);
    }

    getVariable(name: string) {
        return this.variables.get(name);
    }

    hasVariable(name: string) {
        return this.variables.has(name);
    }

    deleteVariable(name: string) {
        this.variables.delete(name);
    }

    addType(name: string, type: Type) {
        this.types.set(name, type);
    }

    getType(name: string) {
        return this.types.get(name);
    }

    hasType(name: string) {
        return this.types.has(name);
    }

    deleteType(name: string) {
        this.types.delete(name);
    }
}

export class CompilerContext {
    expressions: Map<number, Type>;

    constructor() {
        this.expressions = new Map();
    }

    clone() {
        const ctx = new CompilerContext();
        for (const [id, type] of this.expressions) {
            ctx.addExpression(id, type);
        }
        return ctx;
    }

    addExpression(id: number, type: Type) {
        this.expressions.set(id, type);
    }

    getExpression(id: number) {
        return this.expressions.get(id);
    }

    hasExpression(id: number) {
        return this.expressions.has(id);
    }

    deleteExpression(id: number) {
        this.expressions.delete(id);
    }
}
