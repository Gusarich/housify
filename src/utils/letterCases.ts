export function snakeToCamel(s: string) {
    return s.replace(/(_\w)/g, (m) => m[1]!.toUpperCase());
}

export function camelToSnake(s: string) {
    return s.replace(/[A-Z]/g, (m) => '_' + m.toLowerCase()).slice(1);
}
