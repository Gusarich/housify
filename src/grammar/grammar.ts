import { IterationNode } from 'ohm-js';
import {
    AstBinaryOp,
    AstModule,
    AstNode,
    AstUnaryOp,
    createAstNode,
} from './ast';
import grammar from './grammar.ohm-bundle';

function unwrapOptNode(node: IterationNode): IterationNode | undefined {
    return node.children.length === 0 ? undefined : node.children[0];
}

const semantics = grammar.createSemantics();

semantics.addOperation<AstNode>('astOfModule', {
    Module(items) {
        return createAstNode({
            kind: 'module',
            items: items.children.map((item) => item.astOfModuleItem()),
        });
    },
});

semantics.addOperation<AstNode>('astOfModuleItem', {
    ModuleItem(arg0) {
        return arg0.astOfModuleItem();
    },
    House(_arg0, id, _arg2, itemsTop, items, _arg4) {
        return createAstNode({
            kind: 'house',
            name: id.astOfExpression(),
            items: itemsTop.children
                .map((item) => item.astOfHouseItem())
                .concat(items.children.map((item) => item.astOfHouseItem())),
        });
    },
});

semantics.addOperation<AstNode>('astOfHouseItem', {
    HouseItem(arg0) {
        return arg0.astOfHouseItem();
    },
    HouseItemTop(arg0) {
        return arg0.astOfHouseItem();
    },
    GlobalStat(_arg0, name, _arg2, type, _arg4, defaultValueOpt, _arg6) {
        return createAstNode({
            kind: 'globalStat',
            name: name.astOfExpression(),
            type: type.astOfType(),
            defaultValue: unwrapOptNode(defaultValueOpt)?.astOfExpression(),
        });
    },
    PlayerStat(_arg0, name, _arg2, type, _arg4, defaultValueOpt, _arg6) {
        return createAstNode({
            kind: 'playerStat',
            name: name.astOfExpression(),
            type: type.astOfType(),
            defaultValue: unwrapOptNode(defaultValueOpt)?.astOfExpression(),
        });
    },
    Handler(_arg0, _arg1, eventParameter, _arg3, _arg4, statements, _arg6) {
        return createAstNode({
            kind: 'handler',
            event: eventParameter.astOfParameter(),
            statements: statements.children.map((statement) =>
                statement.astOfStatement(),
            ),
        });
    },
});

semantics.addOperation<AstNode>('astOfParameter', {
    Parameter(arg0, _arg1, arg2) {
        return createAstNode({
            kind: 'parameter',
            name: arg0.astOfExpression(),
            type: arg2.astOfType(),
        });
    },
});

semantics.addOperation<AstNode>('astOfType', {
    Type(arg0) {
        return arg0.astOfType();
    },
    Type_regular(arg0) {
        return arg0.astOfExpression();
    },
});

semantics.addOperation<AstNode>('astOfStatement', {
    Statement(arg0) {
        return arg0.astOfStatement();
    },
    StatementAssign(lvalue, _arg1, expression, _arg3) {
        return createAstNode({
            kind: 'statementAssign',
            lvalue: lvalue.astOfExpression(),
            value: expression.astOfExpression(),
        });
    },
    StatementLet(_arg0, name, _arg2, type, _arg4, expression, _arg6) {
        return createAstNode({
            kind: 'statementLet',
            name: name.astOfExpression(),
            type: type.astOfType(),
            value: expression.astOfExpression(),
        });
    },
    StatementExpression(expression, _arg1) {
        return createAstNode({
            kind: 'statementExpression',
            expression: expression.astOfExpression(),
        });
    },
    StatementIf(arg0) {
        return arg0.astOfStatement();
    },
    StatementIf_noElse(_arg0, condition, _arg2, statements, _arg4) {
        return createAstNode({
            kind: 'statementIf',
            condition: condition.astOfExpression(),
            then: statements.children.map((statement) =>
                statement.astOfStatement(),
            ),
        });
    },
    StatementIf_withElse(
        _arg0,
        condition,
        _arg2,
        statements,
        _arg4,
        _arg5,
        _arg6,
        elseStatements,
        _arg8,
    ) {
        return createAstNode({
            kind: 'statementIf',
            condition: condition.astOfExpression(),
            then: statements.children.map((statement) =>
                statement.astOfStatement(),
            ),
            else: elseStatements.children.map((statement) =>
                statement.astOfStatement(),
            ),
        });
    },
    StatementIf_withElseIf(
        _arg0,
        condition,
        _arg2,
        statements,
        _arg4,
        _arg5,
        elseif,
    ) {
        return createAstNode({
            kind: 'statementIf',
            condition: condition.astOfExpression(),
            then: statements.children.map((statement) =>
                statement.astOfStatement(),
            ),
            else: [elseif.astOfStatement()],
        });
    },
});

semantics.addOperation<AstNode>('astOfExpression', {
    Expression(arg0) {
        return arg0.astOfExpression();
    },
    ExpressionBinary(left, op, right) {
        return createAstNode({
            kind: 'expressionBinary',
            op: op.astOfBinaryOp(),
            left: left.astOfExpression(),
            right: right.astOfExpression(),
        });
    },
    ExpressionUnary(op, operand) {
        return createAstNode({
            kind: 'expressionUnary',
            op: op.astOfUnaryOp(),
            operand: operand.astOfExpression(),
        });
    },
    ExpressionLiteral(arg0) {
        return arg0.astOfExpression();
    },
    ExpressionId(arg0) {
        return createAstNode({
            kind: 'expressionId',
            name: arg0.astOfExpression(),
        });
    },
    ExpressionField(struct, _arg1, field) {
        return createAstNode({
            kind: 'expressionField',
            struct: struct.astOfExpression(),
            field: field.astOfExpression(),
        });
    },
    id(_arg0) {
        return createAstNode({
            kind: 'id',
            name: this.sourceString,
        });
    },
    integerLiteral(_arg0) {
        return createAstNode({
            kind: 'integerLiteral',
            value: this.sourceString,
        });
    },
    booleanLiteral(_arg0) {
        return createAstNode({
            kind: 'booleanLiteral',
            value: this.sourceString === 'true',
        });
    },
    ExpressionParens(_arg0, expression, _arg2) {
        return expression.astOfExpression();
    },
});

semantics.addOperation<AstBinaryOp>('astOfBinaryOp', {
    binaryOp(op) {
        switch (op.sourceString) {
            case '+':
            case '-':
            case '*':
            case '/':
            case '==':
            case '!=':
            case '<':
            case '<=':
            case '>':
            case '>=':
            case '&&':
            case '||':
                return op.sourceString as AstBinaryOp;
            default:
                throw new Error(
                    `Unexpected binary operator: '${op.sourceString}'`,
                );
        }
    },
});

semantics.addOperation<AstUnaryOp>('astOfUnaryOp', {
    unaryOp(op) {
        switch (op.sourceString) {
            case '+':
            case '-':
            case '!':
                return op.sourceString as AstUnaryOp;
            default:
                throw new Error(
                    `Unexpected unary operator: '${op.sourceString}'`,
                );
        }
    },
});

export function parse(src: string): AstModule {
    const matchResult = grammar.match(src);
    if (matchResult.failed()) {
        throw Error(matchResult.message);
    }
    return semantics(matchResult).astOfModule();
}
