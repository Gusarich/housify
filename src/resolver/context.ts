import { AstExpression } from '../grammar/ast';
import { EventType } from '../housing/events';
import { Type } from './type';

export type Variable = {
    type: Type;
    constant: boolean;
};

export class StatementContext {
    variables: Map<string, Variable>;

    constructor() {
        this.variables = new Map();
    }

    clone() {
        const sctx = new StatementContext();
        for (const [name, variable] of this.variables) {
            sctx.addVariable(name, variable);
        }
        return sctx;
    }

    addVariable(name: string, type: Variable) {
        this.variables.set(name, type);
    }

    getVariable(name: string) {
        return this.variables.get(name);
    }

    hasVariable(name: string) {
        return this.variables.has(name);
    }

    deleteVariable(name: string) {
        this.variables.delete(name);
    }
}

export type House = {
    globalStats: {
        name: string;
        type: Type;
        defaultValue?: AstExpression;
    }[];
    playerStats: {
        name: string;
        type: Type;
        defaultValue?: AstExpression;
    }[];
    handlers: EventType[];
};

export type StaticConstant = {
    type: Type;
    value: string;
};

export class CompilerContext {
    expressions: Map<number, Type>;
    houses: Map<string, House>;
    types: Map<string, Type>;
    staticConstants: Map<string, StaticConstant>;

    constructor() {
        this.expressions = new Map();
        this.houses = new Map();
        this.types = new Map();
        this.staticConstants = new Map();
    }

    clone() {
        const ctx = new CompilerContext();
        for (const [id, type] of this.expressions) {
            ctx.addExpression(id, type);
        }
        for (const [name, house] of this.houses) {
            ctx.addHouse(name, house);
        }
        for (const [name, type] of this.types) {
            ctx.types.set(name, type);
        }
        for (const [name, constant] of this.staticConstants) {
            ctx.addStaticConstant(name, constant);
        }
        return ctx;
    }

    addExpression(id: number, type: Type) {
        this.expressions.set(id, type);
    }

    getExpression(id: number) {
        return this.expressions.get(id);
    }

    hasExpression(id: number) {
        return this.expressions.has(id);
    }

    deleteExpression(id: number) {
        this.expressions.delete(id);
    }

    addHouse(name: string, house: House) {
        this.houses.set(name, house);
    }

    getHouse(name: string) {
        return this.houses.get(name);
    }

    hasHouse(name: string) {
        return this.houses.has(name);
    }

    deleteHouse(name: string) {
        this.houses.delete(name);
    }

    addType(name: string, type: Type) {
        this.types.set(name, type);
    }

    getType(name: string) {
        return this.types.get(name);
    }

    hasType(name: string) {
        return this.types.has(name);
    }

    deleteType(name: string) {
        this.types.delete(name);
    }

    addStaticConstant(name: string, constant: StaticConstant) {
        this.staticConstants.set(name, constant);
    }

    getStaticConstant(name: string) {
        return this.staticConstants.get(name);
    }

    hasStaticConstant(name: string) {
        return this.staticConstants.has(name);
    }

    deleteStaticConstant(name: string) {
        this.staticConstants.delete(name);
    }
}
