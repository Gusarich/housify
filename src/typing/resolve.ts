import { AstExpression, AstStatement } from '../grammar/ast';
import { StatementContext } from './context';
import { Type } from './type';

export function resolveExpression(
    expression: AstExpression,
    sctx: StatementContext,
): Type {
    switch (expression.kind) {
        case 'expressionId': {
            if (!sctx.hasVariable(expression.name.name)) {
                throw new Error(`Variable '${expression.name.name}' not found`);
            }
            return sctx.getVariable(expression.name.name)!;
        }
        case 'integerLiteral':
            return { name: 'int' };
        case 'expressionBinary': {
            const left = resolveExpression(expression.left, sctx);
            const right = resolveExpression(expression.right, sctx);
            switch (expression.op) {
                case '+':
                case '-':
                case '*':
                case '/':
                case '<':
                case '<=':
                case '>':
                case '>=':
                    if (left.name !== 'int' || right.name !== 'int') {
                        throw new Error('Operands must be integers');
                    }
                    return { name: 'int' };
                case '==':
                case '!=':
                    if (left.name !== right.name) {
                        throw new Error('Operands must be of the same type');
                    }
                    return { name: 'bool' };
                case '&&':
                case '||':
                    if (left.name !== 'bool' || right.name !== 'bool') {
                        throw new Error('Operands must be booleans');
                    }
                    return { name: 'bool' };
            }
        }
        case 'expressionUnary': {
            const operand = resolveExpression(expression.operand, sctx);
            switch (expression.op) {
                case '+':
                case '-':
                    if (operand.name !== 'int') {
                        throw new Error('Operand must be an integer');
                    }
                    return { name: 'int' };
                case '!':
                    if (operand.name !== 'bool') {
                        throw new Error('Operand must be a boolean');
                    }
                    return { name: 'bool' };
            }
        }
        case 'expressionField': {
            const struct = resolveExpression(expression.struct, sctx);
            if (struct.name !== 'struct') {
                throw new Error('Expression must be a struct');
            }
            const field = struct.fields.find(
                (field) => field.name === expression.field.name,
            );
            if (!field) {
                throw new Error(`Field '${expression.field.name}' not found`);
            }
            return field.type;
        }
    }
}

export function resolveStatement(
    statement: AstStatement,
    sctx: StatementContext,
) {
    switch (statement.kind) {
        case 'statementAssign': {
            if (statement.lvalue.kind !== 'expressionId') {
                throw new Error(
                    `'${statement.lvalue.kind}' cannot be assigned`,
                );
            }
            const variableType = sctx.getVariable(statement.lvalue.name.name);
            if (!variableType) {
                throw new Error(
                    `Variable '${statement.lvalue.name.name}' not found`,
                );
            }
            const expressionType = resolveExpression(statement.value, sctx);
            if (variableType.name !== expressionType.name) {
                throw new Error(
                    `Cannot assign '${expressionType.name}' to '${variableType.name}'`,
                );
            }
            break;
        }
        case 'statementIf': {
            const conditionType = resolveExpression(statement.condition, sctx);
            if (conditionType.name !== 'bool') {
                throw new Error('Condition must be a boolean');
            }
            for (const thenStatement of statement.then) {
                resolveStatement(thenStatement, sctx.clone());
            }
            if (statement.else) {
                for (const elseStatement of statement.else) {
                    resolveStatement(elseStatement, sctx.clone());
                }
            }
            break;
        }
        case 'statementLet': {
            if (sctx.hasVariable(statement.name.name)) {
                throw new Error(
                    `Variable '${statement.name.name}' already exists`,
                );
            }
            const valueType = resolveExpression(statement.value, sctx);
            if (statement.type.name !== valueType.name) {
                throw new Error(
                    `Cannot assign '${valueType.name}' to '${statement.type.name}'`,
                );
            }
            sctx.addVariable(statement.name.name, valueType);
            break;
        }
        case 'statementExpression': {
            resolveExpression(statement.expression, sctx);
            break;
        }
    }
}
