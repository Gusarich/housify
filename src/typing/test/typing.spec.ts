import { resolveModule } from '../resolve';
import { loadCases } from '../../utils/loadCases';
import { resetNodeId } from '../../grammar/ast';
import { parse } from '../../grammar/grammar';
import { CompilerContext } from '../context';

describe('Grammar', () => {
    beforeEach(() => {
        resetNodeId();
    });

    for (const r of loadCases(__dirname + '/cases/')) {
        it('should parse ' + r.name, () => {
            const moduleAst = parse(r.code);
            expect(moduleAst).toMatchSnapshot();
            const ctx = new CompilerContext();
            resolveModule(moduleAst, ctx);
            expect(ctx).toMatchSnapshot();
        });
    }

    for (const r of loadCases(__dirname + '/cases-failed/')) {
        it('should fail ' + r.name, () => {
            const moduleAst = parse(r.code);
            const ctx = new CompilerContext();
            expect(() => {
                resolveModule(moduleAst, ctx);
            }).toThrowErrorMatchingSnapshot();
        });
    }
});
