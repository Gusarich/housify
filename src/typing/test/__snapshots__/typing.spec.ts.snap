// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grammar should fail 1 1`] = `"Cannot assign 'int' to 'bool'"`;

exports[`Grammar should parse 1 1`] = `
{
  "id": 24,
  "items": [
    {
      "id": 23,
      "items": [
        {
          "defaultValue": {
            "id": 4,
            "kind": "integerLiteral",
            "value": "10",
          },
          "id": 5,
          "kind": "globalStat",
          "name": {
            "id": 2,
            "kind": "id",
            "name": "x",
          },
          "type": {
            "id": 3,
            "kind": "id",
            "name": "int",
          },
        },
        {
          "event": {
            "id": 6,
            "kind": "id",
            "name": "BlockBreak",
          },
          "id": 22,
          "kind": "handler",
          "statements": [
            {
              "id": 12,
              "kind": "statementAssign",
              "lvalue": {
                "field": {
                  "id": 9,
                  "kind": "id",
                  "name": "x",
                },
                "id": 10,
                "kind": "expressionField",
                "struct": {
                  "id": 8,
                  "kind": "expressionId",
                  "name": {
                    "id": 7,
                    "kind": "id",
                    "name": "global",
                  },
                },
              },
              "value": {
                "id": 11,
                "kind": "integerLiteral",
                "value": "123",
              },
            },
            {
              "id": 21,
              "kind": "statementLet",
              "name": {
                "id": 13,
                "kind": "id",
                "name": "y",
              },
              "type": {
                "id": 14,
                "kind": "id",
                "name": "int",
              },
              "value": {
                "id": 20,
                "kind": "expressionBinary",
                "left": {
                  "field": {
                    "id": 17,
                    "kind": "id",
                    "name": "x",
                  },
                  "id": 18,
                  "kind": "expressionField",
                  "struct": {
                    "id": 16,
                    "kind": "expressionId",
                    "name": {
                      "id": 15,
                      "kind": "id",
                      "name": "global",
                    },
                  },
                },
                "op": "+",
                "right": {
                  "id": 19,
                  "kind": "integerLiteral",
                  "value": "1",
                },
              },
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 1 2`] = `
CompilerContext {
  "expressions": Map {
    4 => {
      "type": "int",
    },
    8 => {
      "fields": [
        {
          "name": "x",
          "type": {
            "type": "int",
          },
        },
      ],
      "name": "global",
      "type": "struct",
    },
    10 => {
      "type": "int",
    },
    11 => {
      "type": "int",
    },
    16 => {
      "fields": [
        {
          "name": "x",
          "type": {
            "type": "int",
          },
        },
      ],
      "name": "global",
      "type": "struct",
    },
    18 => {
      "type": "int",
    },
    19 => {
      "type": "int",
    },
    20 => {
      "type": "int",
    },
  },
}
`;
